app/controllers/accounts_controller.rb

```diff
@@ -609,6 +620,20 @@ def courses_api
       @courses = @courses.not_associated_courses
     end
 
+    if starts_before || ends_after
+      @courses = @courses.joins(:enrollment_term)
+      if starts_before
+        @courses = @courses.where("
+        (courses.start_at IS NULL AND enrollment_terms.start_at IS NULL)
+        OR courses.start_at <= ? OR enrollment_terms.start_at <= ?", starts_before, starts_before)
```

Comment:
isn't it that if the course date is set, that takes precedence over the term date? if so, it should be

`"COALESCE(courses.start_at, enrollment_terms.start_at) <= ?"`

which you could then take to the next step:

`where("COALESCE(COALESCE(courses.start_at, enrollment_terms.start_at) <= ?, ?)", starts_before, true)`

(the result of a comparison between NULL and anything is NULL, so if they're both NULL, the outer COALESCE will do something - return TRUE. if either is NOT NULL, you get the result of the comparison)

Reply:
I don't think I can answer that question for you definitively. I was following the same business logic that's outlined here: https://github.com/instructure/canvas-lms/blob/a664cdb0b26bf9d4473c0204dba38fc73a34ece7/app/controllers/sis_api_controller.rb#L375-L384

However, looking at `Course`'s `:not_completed`, it's a logical `AND` that both the course `conclude_at` and term's `end_at` are less than `now`.

If I interpreted that wrong, I can change this.

Reply:
while I wouldn't trust that controller to be definitive for anything, the not_complete scope on Course does behave as you suggest. it's still a bit weird to me, since enrollment dates (in EnrollmentDateBuilder#build prefers course dates over term dates (if the course is restricted, and the enrollment is not a teacher). But without the context of a specific user in a course, I don't know that we can do much better for saying when a course starts. So... this query is fine as is. Just please add to the documentation on these parameters how they behave to make it clear.

Reply:
If you feel it'd be better for the range to be inclusive of both the course and term's dates I'm fine changing it to that. I don't feel strongly in regards to the AND/OR nature of it or the idea of using `COALESCE`.

Reply:
nah, it's fine as is.

Reply:
I've updated the documentation to explicitly mention that either the course's start_at or the term's start_at must be less than the value or both start_at values are `NULL`. And the same for `end_at`.

I also rebased on the latest version of master.